planner/core/rule_column_pruning.go

```diff
@@ -282,7 +282,9 @@ func (ds *DataSource) PruneColumns(parentUsedCols []*expression.Column) error {
 	originSchemaColumns := ds.schema.Columns
 	originColumns := ds.Columns
 	for i := len(used) - 1; i >= 0; i-- {
-		if !used[i] && !exprUsed[i] {
+		// if ds has a shard index, it can't prune the generated column of shard index
+		// so add the new condtion `!ds.containExprPrefixUk`
+		if !used[i] && !exprUsed[i] && !ds.containExprPrefixUk {
```

Comment:
Does it means that if there is a shard index, no columns will be pruned?

Reply:
yes, it will prune the tidb_shard() gc column!